aicsimageio/writers/ome_tiff_writer.py

```diff
@@ -87,31 +89,66 @@ def save(self, data, ome_xml=None, channel_names=None, image_name="IMAGE0", pixe
         if self.silent_pass:
             return
 
-        tif = tifffile.TiffWriter(self.file_path, bigtiff=self._size_of_ndarray(data=data) > BYTE_BOUNDARY)
-
         shape = data.shape
-        assert (len(shape) == 5 or len(shape) == 4 or len(shape) == 3)
-
-        # if this is 3d data, then assume it's ZYX and transform it to the expected TZCYX
-        if len(shape) == 3:
-            data = np.expand_dims(data, axis=1)
+        ndims = len(shape)
+
+        assert (ndims == 5 or ndims == 4 or ndims == 3), "Expected 3, 4, or 5 dimensions in data array"
+
+        # assert valid characters in dimension_order
+        if not (all(d in "STCZYX" for d in dimension_order)):
+            raise InvalidDimensionOrderingError(f"Invalid dimension_order {dimension_order}")
+        if (dimension_order[-2:] != "YX"):
+            raise InvalidDimensionOrderingError(f"Last two characters of dimension_order {dimension_order} expected to \
+                be YX.  Please transpose your data.")
+        if (len(dimension_order) < ndims):
+            raise InvalidDimensionOrderingError(f"dimension_order {dimension_order} must have at least as many \
+                dimensions as data shape {shape}")
+        if (dimension_order.find("S") > 0):
+            raise InvalidDimensionOrderingError(f"S must be the leading dim in dimension_order {dimension_order}")
+        # todo ensure no letter appears more than once?
+
+        # ensure dimension_order is same len as shape
+        if len(dimension_order) > ndims:
+            dimension_order = dimension_order[-ndims:]
+
+        # if this is 3d data, then expand to 5D and add appropriate dimensions
+        if ndims == 3:
             data = np.expand_dims(data, axis=0)
-        # if this is 4d data, then assume it's ZCYX and transform it to the expected TZCYX
-        elif len(shape) == 4:
             data = np.expand_dims(data, axis=0)
+            # prepend either TC, TZ or CZ
+            if dimension_order[0] == 'T':
+                dimension_order = "CZ" + dimension_order
+            elif dimension_order[0] == 'C':
+                dimension_order = "TZ" + dimension_order
+            elif dimension_order[0] == 'Z':
+                dimension_order = "TC" + dimension_order
+
+        # if this is 4d data, then expand to 5D and add appropriate dimensions
+        elif ndims == 4:
+            data = np.expand_dims(data, axis=0)
+            # prepend either T, C, or Z
+            first2 = dimension_order[:2]
+            if first2 == "TC" or first2 == "CT":
+                dimension_order = "Z" + dimension_order
+            elif first2 == "TZ" or first2 == "ZT":
+                dimension_order = "C" + dimension_order
+            elif first2 == "CZ" or first2 == "ZC":
+                dimension_order = "T" + dimension_order
```

Comment:
Why not use the transform module to expand the dims and reorder? This seems like unnecessary duplication. Part of this is necessary but unless I'm missing it much of this could be replaced with a function call to transform. 
If you use a set, you could clean up the if elif else etc into a simple set subtraction to figure out what to augment for all the cases.

Reply:
Current reasons why I am not using transforms.py:
1. I forgot it was there, and didn't realize it would do any of this. :)
2. I don't think it would save much in the way of readability.
2a. np.expand_dims is a one-liner.
2b. This code is not transposing anything.
3. I would still need to figure out the new dimension ordering before calling the transforms code.